### ISSUE TITLE
How to accelerate reading from hdf5

### ISSUE THREAD

**Issue by seanzhuh** [h5py/h5py#2568]
Hello, I'm just new to h5py package and recently searched the internet but found out there are pretty few useful guide on how to accelerate the reading from hdf5. I know is's a complicated question involving many aspects including the chunk cache size, chunk shape, and compression. Could someone help me to figure out the best hyper-parameters I need to set to obtain the fastest reading speed?

Concretely, I have created a dataset called "images" where the images are of shape 3 x 224 x 224 for the rgb channel, width, and height, and I have 40000 images per .h5 file. In total I have 210000 images, but due to my CPU memory, I can only store at most 40000 images into one .h5 file, leading to 6 .h5 files for those 210000 images. Along with the images I also have binary masks of shape 224 x 224 for each image.

Each image is in numpy.uint8 dtype (1 byte) and each mask is in numpy.bool dtype. And I've set the chunk cache size to be just enough for one chunk.

```
import numpy as np
import h5py

images = np.random.randn(40000, 3, 224, 224).astype(np.uint8)
masks = np.random.randn(40000, 224, 224).astype(np.bool)
h5_file = h5py.File("datasets.h5", "w")
h5_file.create_dataset("images", data=images, 
                       chunks=(12, 3, 224, 224), 
                       compression="gzip", compression_opts=4, 
                       rdcc_nbytes=12*3*224*224, rdcc_nslots=521)
h5_file.create_dataset("masks", data=masks, 
                       chunks=(12, 224, 224), 
                       compression="gzip", compression_opts=4, 
                       rdcc_nbytes=12*224*224, rdcc_nslots=521)
h5_file.close()
```

Each time I will read 12 images and masks, i.e., 
```
h5_file = h5py.File("datasets.h5", "r")
for i in range(40000 // 12): # in practice I randomly shuffled the idxes, but each idx will be called exactly once
    images= h5_file["images"][i * 12: i * 12 + 12]
    masks= h5_file["masks"][i * 12: i * 12 + 12]
```

**Comment by vasole** [comment]
Hello,

Besides reading speed, you also have to consider what are you going to do when processing the images.

Are you going to access the RGB components separately? If not, I would expect to arrange them as (n_frame, n_rows, n_columns, 3) so that you get the RGB components of a pixel in a continuous way and images[0] is the first image that is ready to be displayed/treated applying directly mask[0] at once. If you intend to access the different components separately then your arrangement is more convenient since (0, 0) gives the first Red image but then you will have to apply the mask separately to each component.

**Comment by seanzhuh** [comment]
No I do not access the RGB components separately. The images are input to a neural network defined via pytorch framework. The mask is applied to all rgb channels of the image to mask out some areas to make it blank. Besides the chunk shape, am I doing right on other arguments?

**Comment by ajelenak** [comment]
You can store all your images in a single file (one HDF5 dataset) and you don't need to hold them all in memory for this. For example, you could write out several chunks at one time, similar to what you do when reading images.

Setting chunk cache size to just one chunk is too small when writing out the data. It forces the library to write to storage after just one chunk.  Also, chunk cache size is not stored in the file so you need to set it every time the file is opened for reading. Since you are not setting chunk cache size when reading the data, the default image dataset's chunk cache (1 MiB) is too small for even a single chunk. This may not be a performance problem since it seems there is no need for same chunk data again once processed.

For the current dataset and chunk shapes, there should be about 3,300 chunks per dataset. That's a lot. Perhaps using `libver="latest"` ([docs](https://docs.h5py.org/en/latest/high/file.html#version-bounding)) when creating the files will improve performance.

Better [chunk compression](https://docs.h5py.org/en/latest/high/dataset.html#filter-pipeline) may improve performance but access to more advanced compression methods is not available in h5py. You will need another package. Try [LZF](https://docs.h5py.org/en/latest/high/dataset.html#lossless-compression-filters) and [Shuffle](https://docs.h5py.org/en/latest/high/dataset.html#shuffle-filter) filters since already available in h5py.

**Comment by seanzhuh** [comment]
1) I know that you can append images to a single file even though my CPU memory is not enough to hold them all, but why a single file can accelerate the performance? Since I've built a mapping (dict) for each index that which sub file it will map to, I guess this process is nearly real-time. 

2) I've tried to increase the chunk shape to either (4*12, 3, 224, 224) or (64*12, 3, 224, 224) to reduce the number of chunks per dataset, and accordingly set the chunk cache size to either 4*12*3*224*224 or 64*12*3*224*224. For these two versions, no significant speed was observed, and it is much slower for the 64 version.

3) so when reading the file, I also need to set the chunk cache size right, like this? `h5_file=h5py.File("datasets.j5", "r", rdcc_nbytes=12*3*224*224)`. I forgot to mention that, In the code above, though each index will be read only once, but this is just for one epoch training of neural network. So beyond this inner loop, there is another outer loop that will call each index multiple times, but each inner loop will still read each index only once. It's like read each index of the whole dataset once, then read each index again of the whole dataset, and repeat multiple times.

**Comment by ajelenak** [comment]
> 1. I know that you can append images to a single file even though my CPU memory is not enough to hold them all, but why a single file can accelerate the performance? Since I've built a mapping (dict) for each index that which sub file it will map to, I guess this process is nearly real-time.

Library caching is per file, so open all the files only once and only close them at the very end. Or have all the data in one file.

> 2. I've tried to increase the chunk shape to either (4\*12, 3, 224, 224) or (64\*12, 3, 224, 224) to reduce the number of chunks per dataset, and accordingly set the chunk cache size to either 4\*12\*3\*224\*224 or 64\*12\*3\*224\*224. For these two versions, no significant speed was observed, and it is much slower for the 64 version.

The point is not to always match chunk cache size to only one chunk but to have it large enough for many chunks. Keep the (12, 3, 224, 224) chunk shape but increase the cache to as large size as it is safe for your computer's memory. For example, a cache of 2GB could hold up to ~1,100 chunks at once. Remember that every HDF5 dataset gets its own chunk cache.

> 3. so when reading the file, I also need to set the chunk cache size right, like this? `h5_file=h5py.File("datasets.j5", "r", rdcc_nbytes=12*3*224*224)`.

Yes, and done this way it applies to every opened dataset.

> I forgot to mention that, In the code above, though each index will be read only once, but this is just for one epoch training of neural network. So beyond this inner loop, there is another outer loop that will call each index multiple times, but each inner loop will still read each index only once. It's like read each index of the whole dataset once, then read each index again of the whole dataset, and repeat multiple times.

What is _index_? Is it one _image_, which is  one (1, 3, 224, 224) slice of the image dataset?

**Comment by seanzhuh** [comment]
> > 1. I know that you can append images to a single file even though my CPU memory is not enough to hold them all, but why a single file can accelerate the performance? Since I've built a mapping (dict) for each index that which sub file it will map to, I guess this process is nearly real-time.
> 
> Library caching is per file, so open all the files only once and only close them at the very end. Or have all the data in one file.
> 
> > 2. I've tried to increase the chunk shape to either (4*12, 3, 224, 224) or (64*12, 3, 224, 224) to reduce the number of chunks per dataset, and accordingly set the chunk cache size to either 4*12*3*224*224 or 64*12*3*224*224. For these two versions, no significant speed was observed, and it is much slower for the 64 version.
> 
> The point is not to always match chunk cache size to only one chunk but to have it large enough for many chunks. Keep the (12, 3, 224, 224) chunk shape but increase the cache to as large size as it is safe for your computer's memory. For example, a cache of 2GB could hold up to ~1,100 chunks at once. Remember that every HDF5 dataset gets its own chunk cache.
> 
> > 3. so when reading the file, I also need to set the chunk cache size right, like this? `h5_file=h5py.File("datasets.j5", "r", rdcc_nbytes=12*3*224*224)`.
> 
> Yes, and done this way it applies to every opened dataset.
> 
> > I forgot to mention that, In the code above, though each index will be read only once, but this is just for one epoch training of neural network. So beyond this inner loop, there is another outer loop that will call each index multiple times, but each inner loop will still read each index only once. It's like read each index of the whole dataset once, then read each index again of the whole dataset, and repeat multiple times.
> 
> What is _index_? Is it one _image_, which is one (1, 3, 224, 224) slice of the image dataset?

For the 3rd one, I just add an outer loop compared to the 1st code snippet that I show

h5_file = h5py.File("datasets.h5", "r")
# in practice I randomly shuffled the indexes, and there are 8 processes perform the reading,
# each process will call a different index i, but each index will be called exactly once from exactly one process
# so actually it's parallel reading not shown here.
for i in range(25):
    for j in range(40000 // 12):
        images= h5_file["images"][j * 12: j * 12 + 12]
        masks= h5_file["masks"][j * 12: j * 12 + 12]

**Comment by seanzhuh** [comment]
Also I'd like to ask what is the reading speed for "lzf" and "gzip" compression with compression_opt = 4? Does "lzf" reads faster?

**Comment by seanzhuh** [comment]
I've increased the chunk cache size to 64*1024*12*224*224*3 which is nearly 110GB to hold one chunk with 1.7MB, but the speed is 10 times lower than the chunk cache size 12*224*224*3

**Comment by ajelenak** [comment]
> Also I'd like to ask what is the reading speed for "lzf" and "gzip" compression with compression_opt = 4? Does "lzf" reads faster?

LZF is generaly faster than gzip but you should always test for your data.

**Comment by ajelenak** [comment]
> I've increased the chunk cache size to 64*1024*12*224*224*3 which is nearly 110GB to hold one chunk with 1.7MB, but the speed is 10 times lower than the chunk cache size 12*224*224*3

How large is your computer memory if you opted for a 110 GB chunk cache? And in your case of multiple processes, each one will have this big chunk cache.
